From 11666d6b173ef6af2e0e5350b064d9a21f61f1a9 Mon Sep 17 00:00:00 2001
From: Hetian Fang <hetian.fang@intel.com>
Date: Mon, 18 Aug 2014 17:01:56 +0800
Subject: [PATCH 204/441] USB: dwc3: Implement runtime PM hooks

Re-use system suspend/resume hooks

Note: We must disable the USB PHY for power saving,
before shutdown USB PHY, dwc3_suspend_phy() must be called
to avoid USB hang

Change-Id: Ib3768df35a3850cb2132b7354f55b3e909cd2b91
(cherry picked from commit 092c9f1fa6cb3ce49fb54e43ed8504a499bd1e80)

Conflicts resolved manually in:

	drivers/usb/dwc3/core.c
	drivers/usb/dwc3/gadget.c

Signed-off-by: Brett T. Warden <brett.t.warden@intel.com>
---
 drivers/usb/dwc3/core.c          | 186 ++++++++++++++++++++++++++++-----------
 drivers/usb/dwc3/core.h          |   4 +
 drivers/usb/dwc3/dwc3-pci.c      |   8 ++
 drivers/usb/dwc3/gadget.c        |  23 +++++
 drivers/usb/dwc3/platform_data.h |   1 +
 5 files changed, 173 insertions(+), 49 deletions(-)

--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -62,6 +62,8 @@ void dwc3_set_mode(struct dwc3 *dwc, u32
 #define GUSB2PHYCFG_SUS_PHY                     0x40
 #define GUSB2PHYCFG_PHYSOFTRST (1 << 31)
 #define GUSB2PHYCFG_ULPI_AUTO_RESUME (1 << 15)
+#define GUSB3PIPECTL0                           0xc2c0
+#define GUSB3PIPECTL_SUS_EN                     0x20000
 
 #define EXTEND_ULPI_REGISTER_ACCESS_MASK        0xC0
 #define GUSB2PHYACC0    0xc280
@@ -430,6 +432,60 @@ static void dwc3_core_exit(struct dwc3 *
 	usb_phy_shutdown(dwc->usb3_phy);
 }
 
+static void dwc3_suspend_phy(struct dwc3 *dwc, bool suspend)
+{
+	u32 data = 0;
+
+	data = dwc3_readl(dwc->regs, GUSB2PHYCFG0);
+	if (suspend)
+		data |= GUSB2PHYCFG_SUS_PHY;
+	else
+		data &= ~GUSB2PHYCFG_SUS_PHY;
+
+	dwc3_writel(dwc->regs, GUSB2PHYCFG0, data);
+
+	data = dwc3_readl(dwc->regs, GUSB3PIPECTL0);
+	if (suspend)
+		data |= GUSB3PIPECTL_SUS_EN;
+	else
+		data &= ~GUSB3PIPECTL_SUS_EN;
+
+	dwc3_writel(dwc->regs, GUSB3PIPECTL0, data);
+}
+
+static int dwc3_handle_otg_notification(struct notifier_block *nb,
+		unsigned long event, void *data)
+{
+	struct dwc3* dwc = container_of(nb, struct dwc3, nb);
+	unsigned long flags;
+	int state = NOTIFY_DONE;
+	static int last_value = -1;
+	int val;
+
+	val = *(int *)data;
+
+	if (last_value == val)
+		goto out;
+
+	last_value = val;
+
+	spin_lock_irqsave(&dwc->lock, flags);
+	switch (event) {
+	case USB_EVENT_VBUS:
+		dev_info(dwc->dev, "DWC3 OTG Notify USB_EVENT_VBUS, val = %d\n", val);
+		if (val)
+			pm_runtime_get(dwc->dev);
+		state = NOTIFY_OK;
+		break;
+	default:
+		dev_dbg(dwc->dev, "DWC3 OTG Notify unknow notify message\n");
+	}
+	spin_unlock_irqrestore(&dwc->lock, flags);
+
+out:
+	return state;
+}
+
 #define DWC3_ALIGN_MASK		(16 - 1)
 
 static int dwc3_probe(struct platform_device *pdev)
@@ -485,6 +541,7 @@ static int dwc3_probe(struct platform_de
 
 		dwc->needs_fifo_resize = pdata->tx_fifo_resize;
 		dwc->dr_mode = pdata->dr_mode;
+		dwc->runtime_suspend = pdata->runtime_suspend;
 	} else {
 		dwc->usb2_phy = devm_usb_get_phy(dev, USB_PHY_TYPE_USB2);
 		dwc->usb3_phy = devm_usb_get_phy(dev, USB_PHY_TYPE_USB3);
@@ -554,6 +611,7 @@ static int dwc3_probe(struct platform_de
 	dev->dma_parms	= dev->parent->dma_parms;
 	dma_set_coherent_mask(dev, dev->parent->coherent_dma_mask);
 
+	pm_runtime_set_active(dev);
 	pm_runtime_enable(dev);
 	pm_runtime_get_sync(dev);
 	pm_runtime_forbid(dev);
@@ -629,10 +687,31 @@ static int dwc3_probe(struct platform_de
 		goto err3;
 	}
 
+	atomic_set(&dwc->suspend_depth, 0);
+
+	if (dwc->runtime_suspend) {
+		pm_runtime_set_autosuspend_delay(dev, 500);
+		pm_runtime_use_autosuspend(dev);
+		pm_runtime_mark_last_busy(dev);
+		pm_runtime_put_autosuspend(dev);
+
+		/* Register otg notifier to monitor VBus change events */
+		// FIXME: usb3_phy notification
+		dwc->nb.notifier_call = dwc3_handle_otg_notification;
+		ret = usb_register_notifier(dwc->usb2_phy, &dwc->nb);
+		if (ret) {
+			dev_err(dev, "failed to register otg notifier\n");
+			goto err4;
+		}
+	}
+
 	pm_runtime_allow(dev);
 
 	return 0;
 
+err4:
+	dwc3_debugfs_exit(dwc);
+
 err3:
 	switch (dwc->dr_mode) {
 	case USB_DR_MODE_PERIPHERAL:
@@ -695,64 +774,34 @@ static int dwc3_remove(struct platform_d
 	pm_runtime_put_sync(&pdev->dev);
 	pm_runtime_disable(&pdev->dev);
 
-	return 0;
-}
-
-#ifdef CONFIG_PM_SLEEP
-static int dwc3_prepare(struct device *dev)
-{
-	struct dwc3	*dwc = dev_get_drvdata(dev);
-	unsigned long	flags;
-
-	spin_lock_irqsave(&dwc->lock, flags);
-
-	switch (dwc->dr_mode) {
-	case USB_DR_MODE_PERIPHERAL:
-	case USB_DR_MODE_OTG:
-		dwc3_gadget_prepare(dwc);
-		/* FALLTHROUGH */
-	case USB_DR_MODE_HOST:
-	default:
-		dwc3_event_buffers_cleanup(dwc);
-		break;
+	if (dwc->runtime_suspend) {
+		usb_unregister_notifier(dwc->usb2_phy, &dwc->nb);
+	} else {
+		pm_runtime_put_sync(&pdev->dev);
 	}
 
-	spin_unlock_irqrestore(&dwc->lock, flags);
+	pm_runtime_disable(&pdev->dev);
 
 	return 0;
 }
 
-static void dwc3_complete(struct device *dev)
-{
-	struct dwc3	*dwc = dev_get_drvdata(dev);
-	unsigned long	flags;
-
-	spin_lock_irqsave(&dwc->lock, flags);
-
-	switch (dwc->dr_mode) {
-	case USB_DR_MODE_PERIPHERAL:
-	case USB_DR_MODE_OTG:
-		dwc3_gadget_complete(dwc);
-		/* FALLTHROUGH */
-	case USB_DR_MODE_HOST:
-	default:
-		dwc3_event_buffers_setup(dwc);
-		break;
-	}
 
-	spin_unlock_irqrestore(&dwc->lock, flags);
-}
+#ifdef CONFIG_PM_SLEEP
 
-static int dwc3_suspend(struct device *dev)
+static int dwc3_suspend_common(struct device *dev)
 {
 	struct dwc3	*dwc = dev_get_drvdata(dev);
 	unsigned long	flags;
 
+	if (atomic_inc_return(&dwc->suspend_depth) > 1)
+		return 0;
+
 	spin_lock_irqsave(&dwc->lock, flags);
 
 	switch (dwc->dr_mode) {
 	case USB_DR_MODE_PERIPHERAL:
 	case USB_DR_MODE_OTG:
+		dwc3_gadget_prepare(dwc);
 		dwc3_gadget_suspend(dwc);
 		/* FALLTHROUGH */
 	case USB_DR_MODE_HOST:
@@ -761,7 +810,12 @@ static int dwc3_suspend(struct device *d
 		break;
 	}
 
+	dwc3_event_buffers_cleanup(dwc);
+
 	dwc->gctl = dwc3_readl(dwc->regs, DWC3_GCTL);
+
+	dwc3_suspend_phy(dwc, true);
+
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
 	usb_phy_shutdown(dwc->usb3_phy);
@@ -770,23 +824,30 @@ static int dwc3_suspend(struct device *d
 	return 0;
 }
 
-static int dwc3_resume(struct device *dev)
+static int dwc3_resume_common(struct device *dev)
 {
 	struct dwc3	*dwc = dev_get_drvdata(dev);
 	unsigned long	flags;
 
+	if (atomic_dec_return(&dwc->suspend_depth) > 0)
+		return 0;
+
 	usb_phy_init(dwc->usb3_phy);
 	usb_phy_init(dwc->usb2_phy);
-	msleep(100);
 
 	spin_lock_irqsave(&dwc->lock, flags);
 
+	dwc3_suspend_phy(dwc, false);
+
 	dwc3_writel(dwc->regs, DWC3_GCTL, dwc->gctl);
 
+	dwc3_event_buffers_setup(dwc);
+
 	switch (dwc->dr_mode) {
 	case USB_DR_MODE_PERIPHERAL:
 	case USB_DR_MODE_OTG:
 		dwc3_gadget_resume(dwc);
+		dwc3_gadget_complete(dwc);
 		/* FALLTHROUGH */
 	case USB_DR_MODE_HOST:
 	default:
@@ -796,18 +857,45 @@ static int dwc3_resume(struct device *de
 
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
-	pm_runtime_disable(dev);
-	pm_runtime_set_active(dev);
-	pm_runtime_enable(dev);
+	usb_phy_shutdown(dwc->usb3_phy);
+	usb_phy_shutdown(dwc->usb2_phy);
 
 	return 0;
 }
 
-static const struct dev_pm_ops dwc3_dev_pm_ops = {
-	.prepare	= dwc3_prepare,
-	.complete	= dwc3_complete,
 
+#ifdef CONFIG_PM_RUNTIME
+
+static int dwc3_runtime_suspend(struct device *dev)
+{
+	return dwc3_suspend_common(dev);
+}
+
+static int dwc3_runtime_resume(struct device *dev)
+{
+	return dwc3_resume_common(dev);
+}
+
+#else
+
+#define dwc3_runtime_suspend NULL
+#define dwc3_runtime_resume NULL
+
+#endif
+
+static int dwc3_suspend(struct device *dev)
+{
+	return dwc3_suspend_common(dev);
+}
+
+static int dwc3_resume(struct device *dev)
+{
+	return dwc3_resume_common(dev);
+}
+
+static const struct dev_pm_ops dwc3_dev_pm_ops = {
 	SET_SYSTEM_SLEEP_PM_OPS(dwc3_suspend, dwc3_resume)
+	SET_RUNTIME_PM_OPS(dwc3_runtime_suspend, dwc3_runtime_resume, NULL)
 };
 
 #define DWC3_PM_OPS	&(dwc3_dev_pm_ops)
--- a/drivers/usb/dwc3/core.h
+++ b/drivers/usb/dwc3/core.h
@@ -730,6 +730,10 @@ struct dwc3 {
 
 	u8			test_mode;
 	u8			test_mode_nr;
+
+	bool			runtime_suspend;
+	struct notifier_block	nb;
+	atomic_t		suspend_depth;
 };
 
 /* -------------------------------------------------------------------------- */
--- a/drivers/usb/dwc3/dwc3-pci.c
+++ b/drivers/usb/dwc3/dwc3-pci.c
@@ -146,6 +146,9 @@ static int dwc3_pci_probe(struct pci_dev
 	struct dwc3_pci		*glue;
 	int			ret = -ENOMEM;
 	struct device		*dev = &pci->dev;
+	struct dwc3_platform_data	pdata;
+
+	memset(&pdata, 0x00, sizeof(pdata));
 
 	glue = devm_kzalloc(dev, sizeof(*glue), GFP_KERNEL);
 	if (!glue) {
@@ -176,6 +179,11 @@ static int dwc3_pci_probe(struct pci_dev
 		goto err1;
 	}
 
+	pdata.runtime_suspend = true;
+	ret = platform_device_add_data(dwc3, &pdata, sizeof(pdata));
+	if (ret)
+		goto err1;
+
 	memset(res, 0x00, sizeof(struct resource) * ARRAY_SIZE(res));
 
 	res[0].start	= pci_resource_start(pci, 0);
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -1325,6 +1325,8 @@ static int dwc3_gadget_wakeup(struct usb
 	u8			link_state;
 	u8			speed;
 
+	pm_runtime_get_sync(dwc->dev);
+
 	spin_lock_irqsave(&dwc->lock, flags);
 
 	/*
@@ -1388,6 +1390,8 @@ static int dwc3_gadget_wakeup(struct usb
 out:
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
+	pm_runtime_put_autosuspend(dwc->dev);
+
 	return ret;
 }
 
@@ -1458,10 +1462,14 @@ static int dwc3_gadget_pullup(struct usb
 
 	is_on = !!is_on;
 
+	pm_runtime_get_sync(dwc->dev);
+
 	spin_lock_irqsave(&dwc->lock, flags);
 	ret = dwc3_gadget_run_stop(dwc, is_on);
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
+	pm_runtime_put_autosuspend(dwc->dev);
+
 	return ret;
 }
 
@@ -1521,6 +1529,7 @@ static int dwc3_gadget_start(struct usb_
 		goto err0;
 	}
 
+
 	spin_lock_irqsave(&dwc->lock, flags);
 
 	if (dwc->gadget_driver) {
@@ -1595,8 +1604,11 @@ static int dwc3_gadget_start(struct usb_
 
 	dwc3_gadget_enable_irq(dwc);
 
+
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
+	pm_runtime_put_autosuspend(dwc->dev);
+
 	return 0;
 
 err3:
@@ -1611,6 +1623,8 @@ err1:
 	free_irq(irq, dwc);
 
 err0:
+	pm_runtime_put_autosuspend(dwc->dev);
+
 	return ret;
 }
 
@@ -1621,6 +1635,8 @@ static int dwc3_gadget_stop(struct usb_g
 	unsigned long		flags;
 	int			irq;
 
+	pm_runtime_get_sync(dwc->dev);
+
 	spin_lock_irqsave(&dwc->lock, flags);
 
 	dwc3_gadget_disable_irq(dwc);
@@ -1631,6 +1647,8 @@ static int dwc3_gadget_stop(struct usb_g
 
 	spin_unlock_irqrestore(&dwc->lock, flags);
 
+	pm_runtime_put_autosuspend(dwc->dev);
+
 	irq = platform_get_irq(to_platform_device(dwc->dev), 0);
 	free_irq(irq, dwc);
 
@@ -2156,6 +2174,11 @@ static void dwc3_gadget_disconnect_inter
 
 	dwc->gadget.speed = USB_SPEED_UNKNOWN;
 	dwc->setup_packet_pending = false;
+
+	if (dwc->runtime_suspend) {
+		pm_runtime_mark_last_busy(dwc->dev);
+		pm_runtime_put_autosuspend(dwc->dev);
+	}
 }
 
 static void dwc3_gadget_reset_interrupt(struct dwc3 *dwc)
--- a/drivers/usb/dwc3/platform_data.h
+++ b/drivers/usb/dwc3/platform_data.h
@@ -24,4 +24,5 @@ struct dwc3_platform_data {
 	enum usb_device_speed maximum_speed;
 	enum usb_dr_mode dr_mode;
 	bool tx_fifo_resize;
+	bool runtime_suspend;
 };
